lib/src/tree.rs

```diff
@@ -309,87 +309,99 @@ impl<T> Diff<T> {
     }
 }
 
-struct TreeEntryDiffIterator<'trees, 'matcher> {
-    it1: Peekable<TreeEntriesNonRecursiveIterator<'trees>>,
-    it2: Peekable<TreeEntriesNonRecursiveIterator<'trees>>,
-    // TODO: Restrict walk according to Matcher::visit()
-    _matcher: &'matcher dyn Matcher,
+enum TreeEntryDiffIterator<'trees> {
+    NoWalk,
+    WalkEntries {
+        it1: Peekable<TreeEntriesNonRecursiveIterator<'trees>>,
+        it2: Peekable<TreeEntriesNonRecursiveIterator<'trees>>,
+    },
+    // TODO: Consider adding another variant that walks specific entries given by the matcher
 }
 
-impl<'trees, 'matcher> TreeEntryDiffIterator<'trees, 'matcher> {
-    fn new(tree1: &'trees Tree, tree2: &'trees Tree, matcher: &'matcher dyn Matcher) -> Self {
-        let it1 = tree1.entries_non_recursive().peekable();
-        let it2 = tree2.entries_non_recursive().peekable();
-        TreeEntryDiffIterator {
-            it1,
-            it2,
-            _matcher: matcher,
+impl<'trees> TreeEntryDiffIterator<'trees> {
+    fn new(tree1: &'trees Tree, tree2: &'trees Tree, matcher: &dyn Matcher) -> Self {
+        match matcher.visit(tree1.dir()) {
+            Visit::Nothing => TreeEntryDiffIterator::NoWalk,
+            _ => {
+                let it1 = tree1.entries_non_recursive().peekable();
+                let it2 = tree2.entries_non_recursive().peekable();
+                TreeEntryDiffIterator::WalkEntries { it1, it2 }
+            }
```

Comment:
I don't know how these enum variants will expand, but at this point, I think it's better to leave `TreeEntryDiffIterator::WalkEntries` as a struct (without a matcher), and let the caller decide whether or not to visit the tree pair.

Reply:
I did that as a separate commit. See if you prefer the result and I'll squash them into one. I'm fine with either.

Reply:
Thanks. I prefer the new version as it clarifies `(Maybe)TreeEntryDiffIterator` variants never change during iteration. 
We can also use `Option<TreeEntryDiffIterator>`, but I'm not sure if that's better.

Reply:
I've squashed the two commits. Thanks for reviewing.